### PR TITLE
Add javadoc docstrings to `IndexInterface`, `Index`, `AsyncIndex` classes

### PR DESCRIPTION
**Pull request by austin-denoble**
## Problem
The Java client is lacking docstrings and we're working on adding them.

`IndexInterface`, `Index`, and `AsyncIndex` all need javadoc docstrings with code examples.

## Solution
- Add docstrings to `IndexInterface` and the implementing classes: `Index` and `AsyncIndex`. The code examples are documented in the implementing classes and they inherit the root documentation from `IndexInterface`.

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Infrastructure change (CI configs, etc)
- [X] Non-code change (docs, etc)
- [ ] None of the above: (explain here)

## Test Plan
Pull this branch down and run `gradle generateJavadoc` locally. Even with the errors and warnings you should still get the `/docs` output in the project folder. You can open this in your browser.

Make sure none of the errors or warnings have increased / we're not seeing any outside of the generated code package:
Screenshot 2024-04-10 at 2 46 13 AM

### CODE REVIEW

File: src/main/java/io/pinecone/clients/Index.java [hunk 1]
```diff
@@ -21,6 +25,13 @@ public class Index implements IndexInterface<UpsertResponse,
     private final String indexName;
     private final VectorServiceGrpc.VectorServiceBlockingStub blockingStub;
 
+    /**
+     * Constructs an {@link Index} instance for interacting with a Pinecone index.
+     *
+     * @param connection The {@link PineconeConnection} configuration to be used for this index.
+     * @param indexName The name of the index to interact with. The index host will be automatically resolved.
+     * @throws PineconeValidationException if the connection object is null or the indexName is null or empty.
```
Comment (rohanshah18) [inline src/main/java/io/pinecone/clients/Index.java:33]: index name null or emptiness check is not done at this level

File: src/main/java/io/pinecone/commons/IndexInterface.java [hunk 2]
```diff
@@ -266,67 +377,323 @@ default DescribeIndexStatsRequest validateDescribeIndexStatsRequest(Struct filte
...
+     *
+     * @param topK The number of top similar vectors to retrieve.
+     * @param id The unique ID of a vector to be used as a query vector.
+     * @param namespace The namespace within which the query is performed.
+     * @param filter A Struct defining additional filtering criteria for the query.
+     * @return A generic type {@code U} representing the query results, filtered according to the provided criteria.
+     */
     U queryByVectorId(int topK, String id, String namespace, Struct filter);
 
-    U queryByVectorId(int topK, String id, String namespace, Struct filter, boolean includeValues,
```
Comment (rohanshah18) [inline src/main/java/io/pinecone/commons/IndexInterface.java:288]: Why was this removed?

File: src/main/java/io/pinecone/clients/AsyncIndex.java [hunk 3]
```diff
@@ -24,6 +28,13 @@ public class AsyncIndex implements IndexInterface<ListenableFuture<UpsertRespons
     private final VectorServiceGrpc.VectorServiceFutureStub asyncStub;
     private final String indexName;
 
+    /**
+     * Constructs an {@link AsyncIndex} instance for interacting with a Pinecone index.
+     *
+     * @param connection The {@link PineconeConnection} configuration to be used for this index.
+     * @param indexName The name of the index to interact with. The index host will be automatically resolved.
+     * @throws PineconeValidationException if the connection object is null or the indexName is null or empty.
```
Comment (rohanshah18) [inline src/main/java/io/pinecone/clients/AsyncIndex.java:36]: Same comment as for Index.java, index name is not checked for null or emptiness

**Review by rohanshah18 — APPROVED** [review]
Thanks for pushing these changes, looks good to me. I have left some small comments but LGTM!

File: src/main/java/io/pinecone/clients/AsyncIndex.java [hunk 4]
```diff
@@ -13,6 +13,10 @@
 
 import java.util.List;
 
+/**
+ * A client for interacting with a Pinecone index via GRPC asynchronously. Allows for upserting, querying, fetching, updating, and deleting vectors.
+ * This class provides a direct interface to interact with a specific index, encapsulating network communication and request validation.
+ */
```
Comment (ssmith-pc) [inline src/main/java/io/pinecone/clients/AsyncIndex.java:19]: For "internally" constructed classes like this, I think it's helpful to make it clear to users how to obtain an instance of this class, whether they're expected to construct it themselves or get it via some other means. 

In this case I think you get an an instance of `AsyncIndex` by calling `client.getAsyncIndexConnection("my-index");`

**Review by ssmith-pc — COMMENTED** [review]
FYI this is a lot to go through so rather than holding things up in a single review, I'll just add them as I go so you can act on them sooner.

File: src/main/java/io/pinecone/clients/AsyncIndex.java (hunk 4 above)
Comment (ssmith-pc) [reply src/main/java/io/pinecone/clients/AsyncIndex.java:19]: This could either be placed in this class header portion, or in the constructor javadoc, or both.

File: src/main/java/io/pinecone/clients/AsyncIndex.java [hunk 5]
```diff
@@ -34,26 +45,145 @@ public AsyncIndex(PineconeConnection connection, String indexName) {
         this.asyncStub = connection.getAsyncStub();
     }
 
+    /**
+     * {@inheritDoc}
+     * <pre>{@code
+     *     import io.pinecone.clients.Pinecone;
+     *     import io.pinecone.clients.AsyncIndex;
+     *     import io.pinecone.proto.UpsertResponse;
```
Comment (ssmith-pc) [inline src/main/java/io/pinecone/clients/AsyncIndex.java:53]: @inheritDoc makes this really nice

**Comment by ssmith-pc** [comment]
I'm done with my review -- really nice work @austin-denoble

File: src/main/java/io/pinecone/commons/IndexInterface.java (hunk 2 above)
Comment (austin-denoble) [reply src/main/java/io/pinecone/commons/IndexInterface.java:288]: It wasn't, I think it's just the formatting. It's down on ln. 489.

Comment (rohanshah18) [reply src/main/java/io/pinecone/commons/IndexInterface.java:288]: Ahh okay, sorry
